api/odigos/v1alpha1/collectorsgroup_types.go

```diff
@@ -30,13 +30,51 @@ const (
 	CollectorsGroupRoleNodeCollector  CollectorsGroupRole = CollectorsGroupRole(k8sconsts.CollectorsRoleNodeCollector)
 )
 
+// The raw values of the memory settings for the collectors group.
+// any defaulting, validations and calculations should be done in the controllers
+// that create this CR.
+// Values will be used as is without any further processing.
+type CollectorsGroupMemorySettings struct {
+
+	// MemoryRequestMiB is the memory resource request to be used on the pod template.
+	// it will be embedded in the as a resource request of the form "memory: <value>Mi"
+	MemoryRequestMiB int `json:"memoryRequestMiB"`
+
+	// this parameter sets the "limit_mib" parameter in the memory limiter configuration for the collector.
+	// it is the hard limit after which a force garbage collection will be performed.
+	// this value will end up comparing against the go runtime reported heap Alloc value.
+	// According to the memory limiter docs:
+	// > Note that typically the total memory usage of process will be about 50MiB higher than this value
+	// a test from nov 2024 showed that fresh odigos collector with no traffic takes 38MiB,
+	// thus the 50MiB is a good value to start with.
+	MemoryLimiterLimitMiB int `json:"memoryLimiterLimitMiB"`
+
+	// this parameter sets the "spike_limit_mib" parameter in the memory limiter configuration for the collector memory limiter.
+	// note that this is not the processor soft limit itself, but the diff in Mib between the hard limit and the soft limit.
+	// according to the memory limiter docs, it is recommended to set this to 20% of the hard limit.
+	// changing this value allows trade-offs between memory usage and resiliency to bursts.
+	MemoryLimiterSpikeLimitMiB int `json:"memoryLimiterSpikeLimitMiB"`
+
+	// the GOMEMLIMIT environment variable value for the collector pod.
+	// this is when go runtime will start garbage collection.
+	// it is recommended to be set to 80% of the hard limit of the memory limiter.
+	GomemlimitMiB int `json:"gomemlimitMiB"`
+}
+
 // CollectorsGroupSpec defines the desired state of Collector
 type CollectorsGroupSpec struct {
 	Role CollectorsGroupRole `json:"role"`
 
 	// The port to use for exposing the collector's own metrics as a prometheus endpoint.
 	// This can be used to resolve conflicting ports when a collector is using the host network.
 	CollectorOwnMetricsPort int32 `json:"collectorOwnMetricsPort"`
+
+	// Memory settings for the collectors group.
+	// these settings are used to protect the collectors instances from:
+	// - running out of memory and being killed by the k8s OOM killer
+	// - consuming all available memory on the node which can lead to node instability
+	// - pushing back pressure to the instrumented applications
+	MemorySettings CollectorsGroupMemorySettings `json:"memorySettings"`
```

Comment:
> Should this have omitempty or default values in the kubebuilder level?

that's intentional, since the consumers just use this value as is so it must be set

> I saw that currently the default values are defined in the code. Perhaps defining them in the schema will be better?

I agree, since my focus now is elsewhere, I prefer to do this refactor and leave the current behavior as is. might be addressed in followup PRs